src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -66,12 +41,14 @@ def __init__(
         self,
         ireq,      # type: InstallRequirement
         finder,    # type: PackageFinder
-        preparer,  # type:RequirementPreparer
+        preparer,  # type: RequirementPreparer
+        factory,   # type: Factory
         make_install_req  # type: InstallRequirementProvider
     ):
         # type: (...) -> None
         assert ireq.link is None, "This is a link, not a specifier"
         self._ireq = ireq
+        self._factory = factory
         self._finder = finder
```

Comment:
I was planning to move both `Requirement.find_matches()` and `Candidate.get_dependencies()` into the factory in a follow-up pull request. But we still need to share the preparer with the candidate because of `Candidate.dist`, which makes me unhappy.

Maybe it would be better to keep these methods in `Requirement` and `Candidate`, and use `self._factory.finder`/`preparer` instead. Not sure where to draw the abstraction line.

Reply:
Preparing the candidate to me (via dist) is fundamental, in my view, so I think that does need to remain where it is.

I'll try to draw a diagram that documents what data/methods go into the process where, and are used where. It may help clarify the abstraction. Something like (preparer gets put in here) --> Provider -> creates Requirement -> creates Candidate -> gets Dist (preparer pops up again here), that shows the places the preparer needs to get "smuggled" through so it can get to where it's needed...